cpp/include/raft/core/device_resources_manager.hpp

```diff
@@ -0,0 +1,580 @@
+  };
+
+  // Mutex used to lock access to shared data until after the first
+  // `get_device_resources` call in each thread
+  mutable std::mutex manager_mutex_{};
+  // Indicates whether or not `get_device_resources` has been called by any
+  // host thread
+  bool params_finalized_{};
+  // Container for underlying device resources to be re-used across host
+  // threads for each device
+  std::vector<resource_components> per_device_components_;
+
+  // Return a lock for accessing shared data
+  [[nodiscard]] auto get_lock() const { return std::unique_lock{manager_mutex_}; }
+
+  // Retrieve the underlying resources to be shared across the
+  // application for the indicated device. This method acquires a lock the
+  // first time it is called in each thread for a specific device to ensure that the
+  // underlying resources have been correctly initialized exactly once across
+  // all host threads.
+  auto const& get_device_components(int device_id)
+  {
+    // Each thread maintains an independent list of devices it has
+    // accessed. If it has not marked a device as initialized, it
+    // acquires a lock to initialize it exactly once. This means that each
+    // thread will lock once for a particular device and not proceed until
+    // some thread has actually generated the corresponding device
+    // components
+    thread_local auto initialized_devices = std::vector<int>{};
+    auto iter                             = std::end(per_device_components_);
+    if (std::find(std::begin(initialized_devices), std::end(initialized_devices), device_id) ==
+        std::end(initialized_devices)) {
+      // Only lock if we have not previously accessed this device on this
+      // thread
+      auto lock = get_lock();
+      initialized_devices.push_back(device_id);
+      // If we are building components, do not allow any further changes to
+      // resource parameters.
+      params_finalized_ = true;
+
+      iter = std::find_if(
+        std::begin(per_device_components_),
+        std::end(per_device_components_),
+        [device_id](auto&& components) { return components.get_device_id() == device_id; });
+      if (iter == per_device_components_.end()) {
+        per_device_components_.emplace_back(device_id, params_);
+        iter = std::prev(std::end(per_device_components_));
+      }
+    } else {
+      // If we have previously accessed this device on this thread, we do not
+      // need to lock. We know that this thread already initialized the device
+      // if no other thread had already done so, so we simply retrieve the
+      // components for this device.
+      iter = std::find_if(
+        std::begin(per_device_components_),
+        std::end(per_device_components_),
+        [device_id](auto&& components) { return components.get_device_id() == device_id; });
+    }
+    return *iter;
+  }
+
+  // Thread-safe setter for the number of streams
+  void set_streams_per_device_(std::optional<std::size_t> num_streams)
```

Comment:
How is this context configured? Is this object being mutated often from different threads?

Reply:
> I'm not entirely sure what you're asking.

I'm asking specifically from within the context of the the existing integrations (Redis and Milvus). Just trying to get a feel for the larger picture and how the threads interact there. I was under the impression there was likely some type of setup or initialization step, which would likely be done at the very beginning in a single thread and that the various threads are just being used to create/use their local `device_resources` instances on whatever device theyve been mapped to (or I guess even depending on whatever device happens to be available in the case of load balancing). I understand that it's better to be safe here and I'm not against that at all (especially since we are already locking in `raft::resources`).